{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import phippery\n",
    "from phippery.utils import *\n",
    "\n",
    "import xarray as xr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import os\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import ot\n",
    "\n",
    "from Bio.Align import substitution_matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=dark],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1F1F1F;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 20px 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: '►';\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: '▼';\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: '(';\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: ')';\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: ',';\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt, dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2 {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.Dataset&gt;\n",
       "Dimensions:                      (peptide_id: 24840, peptide_metadata: 6, sample_id: 45, sample_metadata: 31)\n",
       "Coordinates:\n",
       "  * peptide_id                   (peptide_id) int64 0 1 2 ... 24837 24838 24839\n",
       "  * sample_id                    (sample_id) int64 1461 1421 1432 ... 1467 1420\n",
       "  * sample_metadata              (sample_metadata) object &#x27;Cohort&#x27; ... &#x27;visit...\n",
       "  * peptide_metadata             (peptide_metadata) object &#x27;Oligo&#x27; ... &#x27;is_wt&#x27;\n",
       "Data variables:\n",
       "    counts                       (peptide_id, sample_id) int64 0 0 0 ... 71 1270\n",
       "    sample_table                 (sample_id, sample_metadata) object nan ... nan\n",
       "    peptide_table                (peptide_id, peptide_metadata) object &#x27;aggaa...\n",
       "    cpm                          (peptide_id, sample_id) float64 0.0 ... 1.35...\n",
       "    size_factors                 (peptide_id, sample_id) float64 0.0 ... 1.04...\n",
       "    enrichment                   (peptide_id, sample_id) float64 0.1664 ... 8...\n",
       "    neg_binom_mlxp               (peptide_id, sample_id) float64 0.0 ... 3.37\n",
       "    smooth_flank_1_enr_diff_sel  (peptide_id, sample_id) float64 0.4527 ... -...\n",
       "Attributes:\n",
       "    collapsed_variable:  None</pre><div class='xr-wrap' hidden><div class='xr-header'><div class='xr-obj-type'>xarray.Dataset</div></div><ul class='xr-sections'><li class='xr-section-item'><input id='section-243ffcac-0212-49ec-b7f6-805b0ede1dfd' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-243ffcac-0212-49ec-b7f6-805b0ede1dfd' class='xr-section-summary'  title='Expand/collapse section'>Dimensions:</label><div class='xr-section-inline-details'><ul class='xr-dim-list'><li><span class='xr-has-index'>peptide_id</span>: 24840</li><li><span class='xr-has-index'>peptide_metadata</span>: 6</li><li><span class='xr-has-index'>sample_id</span>: 45</li><li><span class='xr-has-index'>sample_metadata</span>: 31</li></ul></div><div class='xr-section-details'></div></li><li class='xr-section-item'><input id='section-28f2629f-abf3-47aa-923a-5e7dd35d9e3a' class='xr-section-summary-in' type='checkbox'  checked><label for='section-28f2629f-abf3-47aa-923a-5e7dd35d9e3a' class='xr-section-summary' >Coordinates: <span>(4)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>peptide_id</span></div><div class='xr-var-dims'>(peptide_id)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 1 2 3 ... 24836 24837 24838 24839</div><input id='attrs-47851d0b-cd24-49cb-9445-ce79d25b3bbe' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-47851d0b-cd24-49cb-9445-ce79d25b3bbe' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e2aa9ef2-9a84-4f5a-b378-11aa4d79ba85' class='xr-var-data-in' type='checkbox'><label for='data-e2aa9ef2-9a84-4f5a-b378-11aa4d79ba85' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([    0,     1,     2, ..., 24837, 24838, 24839])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_id</span></div><div class='xr-var-dims'>(sample_id)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>1461 1421 1432 ... 1466 1467 1420</div><input id='attrs-26f14c09-dded-4c61-bac3-60c30eaf0aac' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-26f14c09-dded-4c61-bac3-60c30eaf0aac' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-85b18a2f-0280-42f0-bd06-08255632fceb' class='xr-var-data-in' type='checkbox'><label for='data-85b18a2f-0280-42f0-bd06-08255632fceb' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([1461, 1421, 1432, 1483, 1463, 1441, 1445, 1447, 1416, 1468, 1422, 1423,\n",
       "       1424, 1474, 1426, 1476, 1477, 1478, 1413, 1430, 1431, 1482, 1435, 1436,\n",
       "       1437, 1438, 1439, 1440, 1491, 1492, 1444, 1446, 1464, 1448, 1449, 1450,\n",
       "       1451, 1452, 1453, 1454, 1455, 1505, 1466, 1467, 1420])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>sample_metadata</span></div><div class='xr-var-dims'>(sample_metadata)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Cohort&#x27; ... &#x27;visit_number&#x27;</div><input id='attrs-50111049-6436-4fb7-a0a6-6a48726cbbc2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-50111049-6436-4fb7-a0a6-6a48726cbbc2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-57d52ac9-aee6-4fe8-9bca-7e908e98b9c8' class='xr-var-data-in' type='checkbox'><label for='data-57d52ac9-aee6-4fe8-9bca-7e908e98b9c8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Cohort&#x27;, &#x27;Vaccine&#x27;, &#x27;age&#x27;, &#x27;age_range&#x27;, &#x27;average quality:&#x27;,\n",
       "       &#x27;control_status&#x27;, &#x27;days_post_infection&#x27;, &#x27;days_post_vaccination&#x27;,\n",
       "       &#x27;dose&#x27;, &#x27;error rate:&#x27;, &#x27;experiment&#x27;, &#x27;fastq_filename&#x27;, &#x27;library_batch&#x27;,\n",
       "       &#x27;manuscript_cohort&#x27;, &#x27;participant_ID&#x27;, &#x27;patient_status_infection&#x27;,\n",
       "       &#x27;patient_status_vaccination&#x27;, &#x27;race&#x27;, &#x27;raw total sequences:&#x27;,\n",
       "       &#x27;reads mapped:&#x27;, &#x27;reference&#x27;, &#x27;sample_ID&#x27;, &#x27;sample_group&#x27;,\n",
       "       &#x27;sample_type&#x27;, &#x27;seq_dir&#x27;, &#x27;sex&#x27;, &#x27;source&#x27;, &#x27;species&#x27;, &#x27;submitted_by&#x27;,\n",
       "       &#x27;technical_replicate_id&#x27;, &#x27;visit_number&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>peptide_metadata</span></div><div class='xr-var-dims'>(peptide_metadata)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;Oligo&#x27; &#x27;Virus&#x27; ... &#x27;is_wt&#x27;</div><input id='attrs-c34511d6-ac80-499e-b95d-55737474175b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-c34511d6-ac80-499e-b95d-55737474175b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-77dc3323-1a67-4279-9473-3d1c082966dd' class='xr-var-data-in' type='checkbox'><label for='data-77dc3323-1a67-4279-9473-3d1c082966dd' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;Oligo&#x27;, &#x27;Virus&#x27;, &#x27;Protein&#x27;, &#x27;Loc&#x27;, &#x27;aa_sub&#x27;, &#x27;is_wt&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-548246f0-a73d-40bc-9297-cc4b7f0bb138' class='xr-section-summary-in' type='checkbox'  checked><label for='section-548246f0-a73d-40bc-9297-cc4b7f0bb138' class='xr-section-summary' >Data variables: <span>(8)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span>counts</span></div><div class='xr-var-dims'>(peptide_id, sample_id)</div><div class='xr-var-dtype'>int64</div><div class='xr-var-preview xr-preview'>0 0 0 0 1 4 ... 295 436 68 71 1270</div><input id='attrs-21a1559b-ab88-427f-980c-13f4bc4ef607' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-21a1559b-ab88-427f-980c-13f4bc4ef607' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-04796827-edfb-400a-b515-4bcc85f3e1ee' class='xr-var-data-in' type='checkbox'><label for='data-04796827-edfb-400a-b515-4bcc85f3e1ee' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[   0,    0,    0, ...,    0,    1,    7],\n",
       "       [   0,    0,    0, ...,    0,    0,    1],\n",
       "       [   0,    0,    1, ...,    0,    1,    2],\n",
       "       ...,\n",
       "       [ 183,   61,  229, ...,  357,  298, 1296],\n",
       "       [  33,   33,  128, ...,  257,  278,  837],\n",
       "       [  22,   73,  487, ...,   68,   71, 1270]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>sample_table</span></div><div class='xr-var-dims'>(sample_id, sample_metadata)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>nan nan nan ... &#x27;caitlin&#x27; 954 nan</div><input id='attrs-6b14d123-2f24-4e80-ab8f-770619fce582' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6b14d123-2f24-4e80-ab8f-770619fce582' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-20e2232e-acb1-4c88-9308-e927948df849' class='xr-var-data-in' type='checkbox'><label for='data-20e2232e-acb1-4c88-9308-e927948df849' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[nan, nan, nan, ..., &#x27;caitlin&#x27;, 995, nan],\n",
       "       [nan, nan, nan, ..., &#x27;caitlin&#x27;, 955, nan],\n",
       "       [nan, nan, nan, ..., &#x27;caitlin&#x27;, 966, nan],\n",
       "       ...,\n",
       "       [nan, nan, nan, ..., &#x27;caitlin&#x27;, 1000, nan],\n",
       "       [nan, nan, nan, ..., &#x27;caitlin&#x27;, 1001, nan],\n",
       "       [nan, nan, nan, ..., &#x27;caitlin&#x27;, 954, nan]], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>peptide_table</span></div><div class='xr-var-dims'>(peptide_id, peptide_metadata)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;aggaattctacgctgagtGGAGGAGGTGGTT...</div><input id='attrs-30aa7ee6-ad82-452d-b50e-aa351d750c9d' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-30aa7ee6-ad82-452d-b50e-aa351d750c9d' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-eef03b6d-9f3f-48b1-8d3a-c8bc65119adc' class='xr-var-data-in' type='checkbox'><label for='data-eef03b6d-9f3f-48b1-8d3a-c8bc65119adc' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[&#x27;aggaattctacgctgagtGGAGGAGGTGGTTCTGGTGGTGGAGGTTCAGGTGGTGGTGGAAGTGGTTTTGTGTTTCTGGTGCTGCTGCCGCTGGTGAGCAGCCAGTGCGTGtgatagcaagcttgcc&#x27;,\n",
       "        &#x27;SARSCoV2&#x27;, &#x27;S1&#x27;, 1, &#x27;G&#x27;, False],\n",
       "       [&#x27;aggaattctacgctgagtGGAGGAGGTGGTTCTGGTGGTGGAGGTTCAGGTGGTGGTGGAAGTGAATTTGTGTTTCTGGTGCTGCTGCCGCTGGTGAGCAGCCAGTGCGTGtgatagcaagcttgcc&#x27;,\n",
       "        &#x27;SARSCoV2&#x27;, &#x27;S1&#x27;, 1, &#x27;E&#x27;, False],\n",
       "       [&#x27;aggaattctacgctgagtGGAGGAGGTGGTTCTGGTGGTGGAGGTTCAGGTGGTGGTGGAAGTGACTTTGTGTTTCTGGTGCTGCTGCCGCTGGTGAGCAGCCAGTGCGTGtgatagcaagcttgcc&#x27;,\n",
       "        &#x27;SARSCoV2&#x27;, &#x27;S1&#x27;, 1, &#x27;D&#x27;, False],\n",
       "       ...,\n",
       "       [&#x27;aggaattctacgctgagtAGCCTGATTGATCTGCAGGAACTGGGCAAGTATGAGCAGTACATCCAGGGAGGAGGTGGTTCTGGTGGTGGAGGTTCAGGTGGTGGTGGAAGTtgatagcaagcttgcc&#x27;,\n",
       "        &#x27;SARSCoV2&#x27;, &#x27;S2&#x27;, 1211, &#x27;Q&#x27;, False],\n",
       "       [&#x27;aggaattctacgctgagtAGCCTGATTGATCTGCAGGAACTGGGCAAGTATGAGCAGTACATCCACGGAGGAGGTGGTTCTGGTGGTGGAGGTTCAGGTGGTGGTGGAAGTtgatagcaagcttgcc&#x27;,\n",
       "        &#x27;SARSCoV2&#x27;, &#x27;S2&#x27;, 1211, &#x27;H&#x27;, False],\n",
       "       [&#x27;aggaattctacgctgagtAGCCTGATTGATCTGCAGGAACTGGGCAAGTATGAGCAGTACATCCCGGGAGGAGGTGGTTCTGGTGGTGGAGGTTCAGGTGGTGGTGGAAGTtgatagcaagcttgcc&#x27;,\n",
       "        &#x27;SARSCoV2&#x27;, &#x27;S2&#x27;, 1211, &#x27;P&#x27;, False]], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>cpm</span></div><div class='xr-var-dims'>(peptide_id, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 ... 68.31 1.351e+03</div><input id='attrs-e8286934-6d98-49c1-bbaf-9e0d0639b579' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e8286934-6d98-49c1-bbaf-9e0d0639b579' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e2b83cff-c68a-46e7-9d52-782b7fc4b852' class='xr-var-data-in' type='checkbox'><label for='data-e2b83cff-c68a-46e7-9d52-782b7fc4b852' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,\n",
       "        9.60000e-01, 7.45000e+00],\n",
       "       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,\n",
       "        0.00000e+00, 1.06000e+00],\n",
       "       [0.00000e+00, 0.00000e+00, 9.90000e-01, ..., 0.00000e+00,\n",
       "        9.60000e-01, 2.13000e+00],\n",
       "       ...,\n",
       "       [1.92160e+02, 7.07800e+01, 2.27140e+02, ..., 5.39200e+02,\n",
       "        2.86690e+02, 1.37891e+03],\n",
       "       [3.46500e+01, 3.82900e+01, 1.26960e+02, ..., 3.88170e+02,\n",
       "        2.67450e+02, 8.90550e+02],\n",
       "       [2.31000e+01, 8.47100e+01, 4.83050e+02, ..., 1.02710e+02,\n",
       "        6.83100e+01, 1.35125e+03]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>size_factors</span></div><div class='xr-var-dims'>(peptide_id, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 ... 41.78 1.048e+03</div><input id='attrs-67c74460-7ccc-43fa-a9dc-107ce99c3c5a' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-67c74460-7ccc-43fa-a9dc-107ce99c3c5a' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-ccb19768-b000-466f-932c-448c0fa9e6f9' class='xr-var-data-in' type='checkbox'><label for='data-ccb19768-b000-466f-932c-448c0fa9e6f9' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,\n",
       "        5.90000e-01, 5.78000e+00],\n",
       "       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,\n",
       "        0.00000e+00, 8.30000e-01],\n",
       "       [0.00000e+00, 0.00000e+00, 8.10000e-01, ..., 0.00000e+00,\n",
       "        5.90000e-01, 1.65000e+00],\n",
       "       ...,\n",
       "       [2.59710e+02, 5.88400e+01, 1.85460e+02, ..., 1.03811e+03,\n",
       "        1.75360e+02, 1.06932e+03],\n",
       "       [4.68300e+01, 3.18300e+01, 1.03660e+02, ..., 7.47320e+02,\n",
       "        1.63590e+02, 6.90600e+02],\n",
       "       [3.12200e+01, 7.04200e+01, 3.94400e+02, ..., 1.97740e+02,\n",
       "        4.17800e+01, 1.04786e+03]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>enrichment</span></div><div class='xr-var-dims'>(peptide_id, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.1664 0.1664 ... 0.4103 8.004</div><input id='attrs-ebf4d9cf-9596-4818-a6ee-0f38f974a1e3' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ebf4d9cf-9596-4818-a6ee-0f38f974a1e3' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-50671d77-c82b-4d79-833b-9ed91923a127' class='xr-var-data-in' type='checkbox'><label for='data-50671d77-c82b-4d79-833b-9ed91923a127' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.16643313, 0.16643345, 0.16643227, ..., 0.1664331 , 0.32620836,\n",
       "        1.40635519],\n",
       "       [0.16948167, 0.169482  , 0.1694808 , ..., 0.16948164, 0.16948137,\n",
       "        0.34913106],\n",
       "       [0.08795143, 0.08795161, 0.17502235, ..., 0.08795141, 0.17238446,\n",
       "        0.27528695],\n",
       "       ...,\n",
       "       [1.24999127, 0.4645091 , 1.47634729, ..., 3.49578137, 1.86171682,\n",
       "        8.92973931],\n",
       "       [0.3601754 , 0.39695158, 1.29278431, ..., 3.93182133, 2.71217059,\n",
       "        9.00737958],\n",
       "       [0.14264988, 0.5073257 , 2.86511401, ..., 0.61386788, 0.41025071,\n",
       "        8.0040464 ]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>neg_binom_mlxp</span></div><div class='xr-var-dims'>(peptide_id, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.0 0.0 0.0 ... 0.8382 0.2683 3.37</div><input id='attrs-248722e5-4e59-4527-a91b-30699b1495d2' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-248722e5-4e59-4527-a91b-30699b1495d2' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-5e104e24-26b9-4bad-8929-9009834821c8' class='xr-var-data-in' type='checkbox'><label for='data-5e104e24-26b9-4bad-8929-9009834821c8' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[0.        , 0.        , 0.        , ..., 0.        , 0.53731528,\n",
       "        1.26535013],\n",
       "       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,\n",
       "        0.59475867],\n",
       "       [0.        , 0.        , 0.55735328, ..., 0.        , 0.55735328,\n",
       "        0.69301118],\n",
       "       ...,\n",
       "       [1.34050574, 0.40677521, 1.01296125, ..., 4.56670603, 0.96781057,\n",
       "        4.69202037],\n",
       "       [0.61077374, 0.46874261, 1.08942708, ..., 5.62317931, 1.55080273,\n",
       "        5.23711263],\n",
       "       [0.22245126, 0.38888933, 1.45950568, ..., 0.8382433 , 0.26831438,\n",
       "        3.36971612]])</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span>smooth_flank_1_enr_diff_sel</span></div><div class='xr-var-dims'>(peptide_id, sample_id)</div><div class='xr-var-dtype'>float64</div><div class='xr-var-preview xr-preview'>0.4527 0.08707 ... -4.141 -0.9426</div><input id='attrs-ada9e6c9-92d2-49bc-918a-4afff5a31d41' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-ada9e6c9-92d2-49bc-918a-4afff5a31d41' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-8cf0b9c2-2ae8-43ac-bf4e-ca2bee3d9df3' class='xr-var-data-in' type='checkbox'><label for='data-8cf0b9c2-2ae8-43ac-bf4e-ca2bee3d9df3' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([[ 4.52742394e-01,  8.70679521e-02, -3.41055336e-01, ...,\n",
       "         8.70677725e-02, -8.94431928e-02,  4.22377249e-01],\n",
       "       [ 4.62527849e-01,  8.89498155e-02, -3.34506364e-01, ...,\n",
       "         8.89496323e-02, -9.73654999e-01, -2.41178925e+00],\n",
       "       [ 1.08894016e-01,  2.09416652e-02, -3.22897925e-01, ...,\n",
       "         2.09416203e-02, -9.50720053e-01, -2.89516548e+00],\n",
       "       ...,\n",
       "       [-2.68748764e+00, -1.11608141e-01, -3.10680366e+00, ...,\n",
       "         1.11468314e+00, -4.44187934e-01,  3.64456113e-01],\n",
       "       [-6.42218738e+00, -2.48051221e-01, -3.72791520e+00, ...,\n",
       "         1.47286671e+00,  4.75343045e-01,  4.67844800e-01],\n",
       "       [-9.20211915e+00, -3.50623226e-02, -5.11212881e-03, ...,\n",
       "        -4.18598889e+00, -4.14060505e+00, -9.42563293e-01]])</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-9ed6d63a-b38f-400b-87b5-8ef859546444' class='xr-section-summary-in' type='checkbox'  checked><label for='section-9ed6d63a-b38f-400b-87b5-8ef859546444' class='xr-section-summary' >Attributes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'><dt><span>collapsed_variable :</span></dt><dd>None</dd></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.Dataset>\n",
       "Dimensions:                      (peptide_id: 24840, peptide_metadata: 6, sample_id: 45, sample_metadata: 31)\n",
       "Coordinates:\n",
       "  * peptide_id                   (peptide_id) int64 0 1 2 ... 24837 24838 24839\n",
       "  * sample_id                    (sample_id) int64 1461 1421 1432 ... 1467 1420\n",
       "  * sample_metadata              (sample_metadata) object 'Cohort' ... 'visit...\n",
       "  * peptide_metadata             (peptide_metadata) object 'Oligo' ... 'is_wt'\n",
       "Data variables:\n",
       "    counts                       (peptide_id, sample_id) int64 0 0 0 ... 71 1270\n",
       "    sample_table                 (sample_id, sample_metadata) object nan ... nan\n",
       "    peptide_table                (peptide_id, peptide_metadata) object 'aggaa...\n",
       "    cpm                          (peptide_id, sample_id) float64 0.0 ... 1.35...\n",
       "    size_factors                 (peptide_id, sample_id) float64 0.0 ... 1.04...\n",
       "    enrichment                   (peptide_id, sample_id) float64 0.1664 ... 8...\n",
       "    neg_binom_mlxp               (peptide_id, sample_id) float64 0.0 ... 3.37\n",
       "    smooth_flank_1_enr_diff_sel  (peptide_id, sample_id) float64 0.4527 ... -...\n",
       "Attributes:\n",
       "    collapsed_variable:  None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix_name = \"BLOSUM62\"                # substitution matrix for amino acid similarity\n",
    "metric = \"smooth_flank_1_enr_diff_sel\"  # scaled differential selection name\n",
    "\n",
    "# epitope region position boundaries\n",
    "epitope_regions = {\n",
    "    \"FP\"  : [ 805,  835],\n",
    "    \"SHH\" : [1135, 1170]\n",
    "}\n",
    "\n",
    "# output directory\n",
    "outdir = 'results'\n",
    "if not os.path.exists(outdir): os.mkdir(outdir)\n",
    "\n",
    "# input file\n",
    "ds = phippery.load(\"LK_DMS_1rep_layered.phip\")\n",
    "ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample IDs for LK individuals\n",
    "lk_infant_sams = sample_id_coordinate_subset(ds,where='sample_group',is_equal_to='LK African Infant').tolist()\n",
    "lk_mother_sams = sample_id_coordinate_subset(ds,where='sample_group',is_equal_to='LK African Mother').tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up cost matrix\n",
    "alphabet_list = list('ARNDCQEGHILKMFPSTWYV')\n",
    "Naa = len(alphabet_list)\n",
    "substitution_matrix = substitution_matrices.load(matrix_name)\n",
    "\n",
    "nthroot=7.\n",
    "maxMij = np.exp(np.max(-substitution_matrix)/nthroot)\n",
    "cost_matrix=[]\n",
    "for aa in alphabet_list:\n",
    "    row = [-x/nthroot for x in substitution_matrix[aa,:][:Naa]]\n",
    "    cost_row = (np.exp(row)).tolist() + [maxMij for i in range(Naa)]\n",
    "    cost_matrix.append(cost_row)\n",
    "\n",
    "for aa in alphabet_list:\n",
    "    row = [-x/nthroot for x in substitution_matrix[aa,:][:Naa]]\n",
    "    cost_row = [maxMij for i in range(Naa)] + (np.exp(row)).tolist()\n",
    "    cost_matrix.append(cost_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve \"histogram\" representation of escape profile at a site, normalized to unit \"area\"\n",
    "# - first 20 bins are magnitude of escape (scaled differential selection < 0) for each amino acid\n",
    "# - last 20 bins are positive scaled differential selection for each amino acid\n",
    "def get_loc_escape_data(\n",
    "    ds,\n",
    "    sid,\n",
    "    loc\n",
    "):\n",
    "    my_ds = ds.loc[\n",
    "                dict(\n",
    "                    peptide_id=peptide_id_coordinate_subset(ds,where='Loc',is_equal_to=loc),\n",
    "                    sample_id=sid\n",
    "                )\n",
    "            ]\n",
    "    \n",
    "    diff_sel = my_ds[metric].to_pandas().to_numpy().flatten()\n",
    "    \n",
    "    my_df = my_ds.peptide_table.loc[:,['aa_sub']].to_pandas()\n",
    "    my_df['diff_sel'] = diff_sel\n",
    "    \n",
    "    esc_data_neg=[]\n",
    "    esc_data_pos=[]\n",
    "    for aa in alphabet_list:\n",
    "        val = my_df[my_df['aa_sub']==aa]['diff_sel'].item()\n",
    "        if val>0:\n",
    "            esc_data_neg.append(0)\n",
    "            esc_data_pos.append(val)\n",
    "        else:\n",
    "            esc_data_neg.append(-val)\n",
    "            esc_data_pos.append(0)\n",
    "    \n",
    "    esc_data = esc_data_neg + esc_data_pos\n",
    "    \n",
    "    if np.sum(esc_data)==0:\n",
    "        return esc_data\n",
    "    else:\n",
    "        return esc_data/np.sum(esc_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute site weight in the comparison of two escape profiles\n",
    "# The weight is the minimum of the relative contribution of that site to the respective escape profile\n",
    "def get_weights(\n",
    "    ds,\n",
    "    sid1,\n",
    "    sid2,\n",
    "    loc_start,\n",
    "    loc_end\n",
    "):   \n",
    "    loc_sums1=[]\n",
    "    loc_sums2=[]\n",
    "    for loc in range(loc_start, loc_end+1):\n",
    "        ds1 = ds.loc[\n",
    "                dict(\n",
    "                    peptide_id=peptide_id_coordinate_subset(ds,where='Loc',is_equal_to=loc),\n",
    "                    sample_id=sid1\n",
    "                    )\n",
    "                ]\n",
    "\n",
    "        diff_sel1 = ds1[metric].to_pandas().to_numpy().flatten()\n",
    "        loc_sums1.append(0)\n",
    "        for val in diff_sel1:\n",
    "            loc_sums1[-1] = loc_sums1[-1] + abs(val)\n",
    "        \n",
    "        ds2 = ds.loc[\n",
    "                dict(\n",
    "                    peptide_id=peptide_id_coordinate_subset(ds,where='Loc',is_equal_to=loc),\n",
    "                    sample_id=sid2\n",
    "                    )\n",
    "                ]\n",
    "        \n",
    "        diff_sel2 = ds2[metric].to_pandas().to_numpy().flatten()\n",
    "        loc_sums2.append(0)\n",
    "        for val in diff_sel2:\n",
    "            loc_sums2[-1] = loc_sums2[-1] + abs(val)\n",
    "    \n",
    "    loc_sums1 = loc_sums1/np.sum(loc_sums1)\n",
    "    loc_sums2 = loc_sums2/np.sum(loc_sums2)\n",
    "        \n",
    "    weights={}\n",
    "    total=0\n",
    "    for i,loc in zip(range(loc_end-loc_start+1), range(loc_start, loc_end+1)):\n",
    "        val = min(loc_sums1[i], loc_sums2[i])\n",
    "        if loc==1136 or loc==1147: val=0\n",
    "        total = total+val\n",
    "        weights[loc] = val\n",
    "    \n",
    "    weights = {k: v/total for k,v in weights.items()}\n",
    "\n",
    "    return weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute epitope region similarity between two escape profiles\n",
    "def region_compare(ds, sid1, sid2, loc_start, loc_end):\n",
    "    weights = get_weights(ds, sid1, sid2, loc_start, loc_end)\n",
    "    region_sim=0\n",
    "    for loc in range(loc_start, loc_end+1):\n",
    "        a    = get_loc_escape_data(ds,sid1,loc)\n",
    "        b    = get_loc_escape_data(ds,sid2,loc)\n",
    "        cost = ot.emd2(a, b, cost_matrix)\n",
    "        sim  = weights[loc]/cost\n",
    "        if (np.sum(a)==0 and np.sum(b)>0) or (np.sum(a)>0 and np.sum(b)==0):\n",
    "            sim = 0\n",
    "        region_sim = region_sim + sim\n",
    "        \n",
    "    return region_sim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FP LK_Infant LK_Infant\n",
      "SHH LK_Infant LK_Infant\n",
      "FP LK_Infant LK_Mother\n",
      "SHH LK_Infant LK_Mother\n",
      "FP LK_Mother LK_Mother\n",
      "SHH LK_Mother LK_Mother\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>part_ID_1</th>\n",
       "      <th>group_1</th>\n",
       "      <th>part_ID_2</th>\n",
       "      <th>group_2</th>\n",
       "      <th>epitope_region</th>\n",
       "      <th>similarity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ptnum_8</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>ptnum_142</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>FP</td>\n",
       "      <td>1.374935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ptnum_8</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>ptnum_18</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>FP</td>\n",
       "      <td>1.371606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ptnum_8</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>ptnum_41</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>FP</td>\n",
       "      <td>1.425037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ptnum_50</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>ptnum_8</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>FP</td>\n",
       "      <td>1.434415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ptnum_50</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>ptnum_133</td>\n",
       "      <td>LK_Infant</td>\n",
       "      <td>FP</td>\n",
       "      <td>1.551646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1975</th>\n",
       "      <td>ptnum_41</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>ptnum_203</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>SHH</td>\n",
       "      <td>0.927252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1976</th>\n",
       "      <td>ptnum_41</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>ptnum_204</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>SHH</td>\n",
       "      <td>0.985854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1977</th>\n",
       "      <td>ptnum_41</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>ptnum_205</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>SHH</td>\n",
       "      <td>0.899583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1978</th>\n",
       "      <td>ptnum_41</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>ptnum_25</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>SHH</td>\n",
       "      <td>1.092964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1979</th>\n",
       "      <td>ptnum_41</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>ptnum_27</td>\n",
       "      <td>LK_Mother</td>\n",
       "      <td>SHH</td>\n",
       "      <td>0.987084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1980 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     part_ID_1    group_1  part_ID_2    group_2 epitope_region  similarity\n",
       "0      ptnum_8  LK_Infant  ptnum_142  LK_Infant             FP    1.374935\n",
       "1      ptnum_8  LK_Infant   ptnum_18  LK_Infant             FP    1.371606\n",
       "2      ptnum_8  LK_Infant   ptnum_41  LK_Infant             FP    1.425037\n",
       "3     ptnum_50  LK_Infant    ptnum_8  LK_Infant             FP    1.434415\n",
       "4     ptnum_50  LK_Infant  ptnum_133  LK_Infant             FP    1.551646\n",
       "...        ...        ...        ...        ...            ...         ...\n",
       "1975  ptnum_41  LK_Mother  ptnum_203  LK_Mother            SHH    0.927252\n",
       "1976  ptnum_41  LK_Mother  ptnum_204  LK_Mother            SHH    0.985854\n",
       "1977  ptnum_41  LK_Mother  ptnum_205  LK_Mother            SHH    0.899583\n",
       "1978  ptnum_41  LK_Mother   ptnum_25  LK_Mother            SHH    1.092964\n",
       "1979  ptnum_41  LK_Mother   ptnum_27  LK_Mother            SHH    0.987084\n",
       "\n",
       "[1980 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_df = pd.DataFrame(columns=['part_ID_1','group_1','part_ID_2','group_2','epitope_region','similarity'])\n",
    "\n",
    "group_list = [lk_infant_sams, lk_mother_sams]\n",
    "desc_list  = ['LK_Infant', 'LK_Mother']\n",
    "\n",
    "other_group_list = copy.deepcopy(group_list)\n",
    "other_desc_list  = copy.deepcopy(desc_list)\n",
    "for group1, desc1 in zip(group_list, desc_list):\n",
    "    for group2, desc2 in zip(other_group_list, other_desc_list):\n",
    "        for region in ['FP', 'SHH']:\n",
    "            print(region,desc1,desc2)\n",
    "            for sid1 in group1:\n",
    "                part_id1 = ds.loc[dict(sample_id=sid1)].sample_table.loc['participant_ID'].values\n",
    "                for sid2 in group2:\n",
    "                    if desc1==desc2 and sid1>=sid2: continue\n",
    "                    part_id2 = ds.loc[dict(sample_id=sid2)].sample_table.loc['participant_ID'].values\n",
    "                    sim = region_compare(ds, sid1, sid2, epitope_regions[region][0], epitope_regions[region][1])\n",
    "                    output_df.loc[len(output_df.index)] = [part_id1, desc1, part_id2, desc2, region, sim]\n",
    "    other_group_list.remove(group1)\n",
    "    other_desc_list.remove(desc1)\n",
    "\n",
    "output_df.to_csv(f'{outdir}/LK_escape_compare_sims.csv', index=False, na_rep=\"NA\")\n",
    "output_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mother-Mother pairs in FP: 595\n",
      "Mother-Mother pairs in SHH: 595\n"
     ]
    }
   ],
   "source": [
    "for region in ['FP','SHH']:\n",
    "    print(f'Mother-Mother pairs in {region}:',\n",
    "          output_df[(output_df['group_1']=='LK_Mother') & (output_df['group_2']=='LK_Mother') & (output_df['epitope_region']==region)].shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Infant-Infant pairs in FP: 45\n",
      "Infant-Infant pairs in SHH: 45\n"
     ]
    }
   ],
   "source": [
    "for region in ['FP','SHH']:\n",
    "    print(f'Infant-Infant pairs in {region}:',\n",
    "          output_df[(output_df['group_1']=='LK_Infant') & (output_df['group_2']=='LK_Infant') & (output_df['epitope_region']==region)].shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Infant-Mother pairs in FP: 350\n",
      "Infant-Mother pairs in SHH: 350\n"
     ]
    }
   ],
   "source": [
    "for region in ['FP','SHH']:\n",
    "    print(f'Infant-Mother pairs in {region}:',\n",
    "          output_df[(output_df['group_1']=='LK_Infant') & (output_df['group_2']=='LK_Mother') & (output_df['epitope_region']==region)].shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matched Infant-Mother pairs in FP: 9\n",
      "Matched Infant-Mother pairs in SHH: 9\n"
     ]
    }
   ],
   "source": [
    "for region in ['FP','SHH']:\n",
    "    print(f'Matched Infant-Mother pairs in {region}:',\n",
    "          output_df[(output_df['group_1']=='LK_Infant') & (output_df['group_2']=='LK_Mother') & (output_df['part_ID_1']==output_df['part_ID_2']) & (output_df['epitope_region']==region)].shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
